### PR TITLE
Added LuaBuilder plugin support

### PR DESCRIPTION
**Pull request by britzl**
This PR makes it possible for developers to add their own LuaBuilder plugins using the extension system. These builder plugins can be used to process the Lua source code in a number of different ways:

* Minify and/or obfuscate code
* Preprocess code using project settings
* Run a templating engine on the code

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 1]
```diff
@@ -62,6 +65,36 @@
...
+    private static List<LuaBuilderPlugin> luaBuilderPlugins = null;
+
+    private void initBuilderPlugins() throws CompileExceptionError{
+        if (luaBuilderPlugins != null) {
+            return;
+        }
+
+        luaBuilderPlugins = new ArrayList<LuaBuilderPlugin>();
+        IClassScanner scanner = Project.getClassLoaderScanner();
+        Set<String> classNames = scanner.scan("com.dynamo.bob.pipeline");
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:77]: Currently a requirement that any plugins are in the com.dynamo.bob.pipeline package

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 2]
```diff
@@ -118,8 +148,17 @@
             // If a script error occurs in runtime we want Lua to report the end of the filepath
             // associated with the chunk, since this is where the filename is visible.
             //
-            String chunkName = "@" + task.input(0).getPath();
-            ProcessBuilder pb = new ProcessBuilder(new String[] { Bob.getExe(Platform.getHostPlatform(), luajitExe), "-bgf", chunkName, inputFile.getAbsolutePath(), outputFile.getAbsolutePath() }).redirectErrorStream(true);
+            final String chunkName = "@" + task.input(0).getPath();
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:151]: Easier to read and extend when options are split like this

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java [hunk 3]
```diff
@@ -27,9 +35,47 @@
...
+
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.antlr.v4.runtime.tree.ParseTreeWalker;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
+
+public class LuaScanner extends LuaParserBaseListener {
```
Comment (britzl) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java:50]: I've merged LuaScanner and DefoldLuaParser into a single file

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaScannerTest.java [hunk 4]
```diff
@@ -147,7 +156,9 @@ public void testProps() throws Exception {
     @Test
     public void testPropsStripped() throws Exception {
         String source = getFile("test_props.lua");
-        List<Property> properties = LuaScanner.scanProperties(source);
+        LuaScanner scanner = new LuaScanner();
+        source = scanner.parse(source);
+        List<Property> properties = scanner.getProperties();
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/pipeline/LuaScannerTest.java:161]: getPropertiesFromString() ?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 5]
```diff
@@ -184,37 +198,45 @@
...
-        // disabling bytecode generation.
-        byte[] bytecode = constructBytecode(task, task.input(0).getPath(), scriptBytesStripped);
-        if (bytecode != null)
-            srcBuilder.setBytecode(ByteString.copyFrom(bytecode));
-        */
 
-        srcBuilder.setFilename(task.input(0).getPath());
+        // create any builder plugins
+        if (luaBuilderPlugins == null) {
+            luaBuilderPlugins = PluginScanner.createPlugins("com.dynamo.bob.pipeline", LuaBuilderPlugin.class);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:226]: Plural?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 6]
```diff
@@ -184,37 +198,45 @@
...
+        }
+        // apply any builder plugins
+        for(LuaBuilderPlugin builderPlugin : luaBuilderPlugins) {
+            try {
+                script = builderPlugin.build(script);
+            }
+            catch(Exception e) {
+                throw new CompileExceptionError("Unable to run Lua builder plugin", e);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:234]: Pass the resource as well!

Comment (JCash) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:234]: Also, is there a line number?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java [hunk 7]
```diff
@@ -27,9 +35,47 @@
...
+import com.dynamo.bob.pipeline.antlr.LuaLexer;
+import com.dynamo.bob.pipeline.antlr.LuaParserBaseListener;
+
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.antlr.v4.runtime.tree.ParseTreeWalker;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.Token;
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java:48]: Why do we have some code in an external jar, and some code in a local (copied) package (now renamed to antlr)?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/antlr/LuaLexer.java [hunk 8]
```diff
@@ -12,7 +12,7 @@
 // CONDITIONS OF ANY KIND, either express or implied. See the License for the
 // specific language governing permissions and limitations under the License.
 
-package com.dynamo.bob.pipeline.luaparser;
+package com.dynamo.bob.pipeline.antlr;
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/antlr/LuaLexer.java:15]: Can we not use the external package/jar that we're using? (see comment above)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 9]
```diff
@@ -184,37 +198,45 @@
...
 
-        srcBuilder.setFilename(task.input(0).getPath());
+        // create any builder plugins
+        if (luaBuilderPlugins == null) {
+            luaBuilderPlugins = PluginScanner.createPlugins("com.dynamo.bob.pipeline", LuaBuilderPlugin.class);
+        }
+        // apply any builder plugins
+        for(LuaBuilderPlugin builderPlugin : luaBuilderPlugins) {
+            try {
+                script = builderPlugin.build(script);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:231]: How is the order controlled?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/plugin/PluginScanner.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	 * 
+	 * - Not be private
+	 * - Not be abstract
+	 * - Extend the base class
+	 * 
+	 * @param packageName 
+	 * @param pluginBaseClass
+	 * @return List of instances
+	 */
+	public static <T> List<T> createPlugins(String packageName, Class<T> pluginBaseClass) throws CompileExceptionError {
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/plugin/PluginScanner.java:42]: I think the creation of multiple instances is a bit confusing.
We currently run both compilers, and have no control of the order.

**Review by JCash — CHANGES_REQUESTED** [review]
Looks good, but I think there's room for improvement (or clarifications on how things work)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java (hunk 5 above)
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java]: Change to only accept and return a single plugin.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java (hunk 7 above)
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaScanner.java:48]: Antlr is the generic tool/framework we use to parse the Lua code (https://www.antlr.org/). It contains one part that is the parser (the jar) and one part which is the generated parser of Lua code (the java files). This is the way Antlr works by generating a parser based on a language definition. The grammar of the language is defined in the .g4 files in the same package path. This means that Antlr can work on any structured language as long as there is a language definition for it. And there is a huge list of language grammars on GitHub: https://github.com/antlr/grammars-v4

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/antlr/LuaLexer.java (hunk 8 above)
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/antlr/LuaLexer.java:15]: No. See comment above.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java (hunk 9 above)
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java]: As discussed I've now changed to only accept a single plugin. In the future we can add support for multiple plugins and control the order through either an annotation or a function in the base class.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/plugin/PluginScanner.java (hunk 10 above)
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/plugin/PluginScanner.java]: Yep. Changed to a single one.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 11]
```diff
@@ -221,14 +221,11 @@ public void build(Task<Void> task) throws CompileExceptionError, IOException {
...
         builder.addAllPropertyResources(propertyResources);
 
-        // create any builder plugins
-        if (luaBuilderPlugins == null) {
-            luaBuilderPlugins = PluginScanner.createPlugins("com.dynamo.bob.pipeline", LuaBuilderPlugin.class);
-        }
-        // apply any builder plugins
-        for(LuaBuilderPlugin builderPlugin : luaBuilderPlugins) {
+        // create and apply a builder plugin if one exists
+        LuaBuilderPlugin luaBuilderPlugin = PluginScanner.createPlugin("com.dynamo.bob.pipeline", LuaBuilderPlugin.class);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:225]: Why recreate it for every script?
You have the static variable, I assume you intended to use that one?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/plugin/PluginScanner.java [hunk 12]
```diff
@@ -60,6 +76,9 @@
 				if (pluginBaseClass.isAssignableFrom(klass) && !isAbstract && !isPrivate) {
 					Bob.verbose("Found plugin " + className);
 					plugins.add((T)klass.newInstance());
+					if (plugins.size() > 1) {
+						throw new CompileExceptionError("PluginScanner found more than one class implementing " + pluginBaseClass);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/plugin/PluginScanner.java:80]: Perhaps output the package name as well?

Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/plugin/PluginScanner.java:80]: In general, I think it's better form to check if we should return first, before creating more resources.
I think it applies here as well, why not check if the list already has one item and then throw?

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java (hunk 11 above)
Comment (JCash) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:225]: Edit: I now see that you've created a cache inside the create function.
I think the name should be then called `createOrGet()` to communicate that.

**Review by JCash — CHANGES_REQUESTED** [review]

**Review by JCash — APPROVED** [review]

**Pull request merged**